### PR TITLE
refactor!: Update SourceLink concept (EDM rework part 3a/4)

### LINKED ISSUE #463: Evolve SourceLink to handle Surface and Volume measurements
This has already been brought up by @FabianKlimpel once: if we want to include free measurements that are not bound to a surface, then the current `SurfaceLinkConcept` is not sufficient/too restrict.

Requiring a `const Surface& referenceSurface()` is certainly not suitable for volume based measurements, unless we create a special `PositionSurface`, but this will be difficult to fit into the `Surface` base class.

One related question, why do we require the following?
```c++
  constexpr static bool default_constructible =
      std::is_default_constructible_v<T>;
  static_assert(default_constructible,
                "Source link must be default-constructible");
```

That sounds counter-intuitive to requiring a `const Surface& referenceSurface() const` in the Concept.

*Summary of the discussion* (so far):

The `SourceLink` would be changed from requiring a `referenceSurface()` method to providing the `GeometryIdentifier`.

This would allow:
 - identify if the source-linked geometry object is a volume or a surface 
 - keep one consistent requirement between the two measurement types
 - have a primitive type 

At the same point the TrackingGeometry should get a fast access to Surfaces or Volumes given a GeometryIdentifier object.

### PR DESCRIPTION
**Pull request by msmk0**
BREAKING_CHANGE: Types implementing the `SourceLinkConcept` must provide a `.geometryId()` member function instead of a `.referenceSurface()` member function. The `MinimalSourceLink` implementation is now available through a separate `MinimalSourceLink` header file.

With this change a source link can reference any part of the geometry hierarchy including surfaces and volumes. The source link lookup in the `KalmanFitter` and in the `CombinatorialKalmanFilter` now use the geometry identifier instead of the value of the surface pointer. Plugins and examples are updated accordingly.

Requires #510 and #510 and must be merged afterwards. Closes #463.

### CODE REVIEW

File: Tests/UnitTests/Core/TrackFinding/CombinatorialKalmanFilterTests.cpp [hunk 1]
```diff
@@ -55,8 +55,8 @@ struct ExtendedMinimalSourceLink {
       size_t sid, const FittableMeasurement<ExtendedMinimalSourceLink>* m)
       : sourceID(sid), meas(m) {}
 
-  const Surface& referenceSurface() const {
-    return *MeasurementHelpers::getSurface(*meas);
+  GeometryIdentifier geometryId() const {
+    return MeasurementHelpers::getSurface(*meas)->geometryId();
```
Comment (XiaocongAi) [inline Tests/UnitTests/Core/TrackFinding/CombinatorialKalmanFilterTests.cpp:59]: Do we need to explicitly define the equality operators for the `ExtendedMinimalSourceLink` as well?

**Review by XiaocongAi — APPROVED** [review]
Hi @msmk0 , it looks good to me with only a minor comment. Approve.

File: Tests/UnitTests/Core/TrackFinding/CombinatorialKalmanFilterTests.cpp (hunk 1 above)
Comment (msmk0) [reply Tests/UnitTests/Core/TrackFinding/CombinatorialKalmanFilterTests.cpp:59]: Yes, we do. The `ExtendedMinimalSourceLink` is a separate type from the `MinimalSourceLink` in the core. The will be unified into a single `TestSourceLink` in #533.

**Review by asalzburger — APPROVED** [review]
I reviewed most of this as part of #533 which builds on top of that.

**Pull request merged**